File: Java/2477. Minimum Fuel Cost to Report to the Capital.java
class Solution {
    private Map<Integer, List<Integer>> adjs;
    private long ans;

    public long minimumFuelCost(int[][] roads, int seats) {
        adjs = new HashMap();
        ans = 0;

        for (int[] r: roads) {
            adjs.computeIfAbsent(r[0], value -> new ArrayList()).add(r[1]);
            adjs.computeIfAbsent(r[1], value -> new ArrayList()).add(r[0]);
        }

        dfs(roads, seats, 0, -1);
        return ans;
    }

    private long dfs(int[][] roads, int seats, int node, int parent) {
        long rep = 1;

        if (adjs.containsKey(node)) {
            for (int a: adjs.get(node)) {
                if (a != parent) {
                    rep += dfs(roads, seats, a, node);
                }
            }
        }
        
        if (node != 0) {
            ans += Math.ceil((double)rep / seats);
        }

        return rep;
    }
}

File: Python3/2477. Minimum Fuel Cost to Report to the Capital.py
class Solution:
    def minimumFuelCost(self, roads: List[List[int]], seats: int) -> int:
        adjs = defaultdict(list)

        for u, v in roads:
            adjs[u].append(v)
            adjs[v].append(u)

        fuel = 0

        def dfs(node = 0, parent = -1):
            rep = 1
            for a in adjs[node]:
                if a != parent:
                    rep += dfs(a, node)

            nonlocal fuel
            if node != 0:
                fuel += math.ceil(rep/seats)
            return rep

        dfs()
        return fuel